### PR TITLE
Allow multiple arguments for "flux suspend/resume"

### LINKED ISSUE #3746: Multiple arguments for `flux suspend/resume ks`
### Describe the bug

# What
This is a proposal for an enhancement to run `suspend/resume` on multiple Kustomizations using
```console
flux suspend ks operator database app
```

# Why 
Currently, to run `flux suspend/resume ks` on multiple Kustomizations, I need to do the following: 
```console
flux suspend ks operator && \
flux suspend ks database && \
flux suspend ks app
```

This will improve the developer workflow and user experience as I have done the above workaround many times to run the above-mentioned operations on Kustomizations frequently.

### Steps to reproduce

N/A

### Expected behavior

```console
# flux suspend ks operator database app
► resuming kustomization operator in flux-system namespace
✔ kustomization resumed
◎ waiting for Kustomization reconciliation
✔ Kustomization reconciliation completed
✔ applied revision main/<commit-sha>
► resuming kustomization datastores in flux-system namespace
✔ kustomization resumed
◎ waiting for Kustomization reconciliation
✔ Kustomization reconciliation completed
✔ applied revision main/<commit-sha>
► resuming kustomization app in flux-system namespace
✔ kustomization resumed
◎ waiting for Kustomization reconciliation
✔ Kustomization reconciliation completed
✔ applied revision main/<commit-sha>
```

### Screenshots and recordings

_No response_

### OS / Distro

Linux

### Flux version

N/A

### Flux check

N/A

### Git provider

_No response_

### Container Registry provider

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #3793: Improve the way `--wait` reconciles multiple resources
### Describe the bug

Follow-up to #3746 

Currently, when we run:
```console
flux reconcile ks --all --wait
```
what happens is, 
- `suspend: true` is set for the first Kustomization
- it is reconciled
- `suspend: true` is set for the next Kustomization
- it is reconciled
- and so on and so forth.....

This patch-wait-patch-wait behaviour is not consistent with other tools like `kubectl` where the patching is separated from the waiting, so that we can have a patching phase and a waiting phase where the former would patch all objects and only then we wait for all of them to be reconciled, **thus making the overall process faster**.

### Steps to reproduce

N/A

### Expected behavior

N/A

### Screenshots and recordings

_No response_

### OS / Distro

N/A

### Flux version

N/A

### Flux check

N/A

### Git provider

_No response_

### Container Registry provider

_No response_

### Additional context

This was discussed in [this prior pull request](https://github.com/fluxcd/flux2/pull/3755#pullrequestreview-1378748396) and this issue is being created to track the work needed for the same.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by rishinair11**
This PR adds support for running `suspend/resume` on multiple supported resources at the same time. This improves the user experience by converting

```console
flux suspend/resume ks operator && \
flux suspend/resume ks database && \
flux suspend/resume ks app
```
to

```console
flux suspend/resume ks operator database app
```

This works for all types of resources (Kustomizations, Sources, etc.) since it has been implemented at the `suspend.go` and `resume.go` level.

This PR also runs resource reconciliations concurrently using go-routines.

Command output:
```console
# flux resume kustomization tkfg tkfg2 tkfg3 tkfg4 tkfg5 tkfg6 --wait
► resuming kustomization tkfg in flux-system namespace
✔ kustomization resumed
► resuming kustomization tkfg2 in flux-system namespace
✔ kustomization resumed
► resuming kustomization tkfg3 in flux-system namespace
✔ kustomization resumed
► resuming kustomization tkfg4 in flux-system namespace
✔ kustomization resumed
► resuming kustomization tkfg5 in flux-system namespace
✔ kustomization resumed
► resuming kustomization tkfg6 in flux-system namespace
✔ kustomization resumed
◎ waiting for Kustomization reconciliation
◎ waiting for Kustomization reconciliation
◎ waiting for Kustomization reconciliation
◎ waiting for Kustomization reconciliation
◎ waiting for Kustomization reconciliation
◎ waiting for Kustomization reconciliation
✔ Kustomization tkfg reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
✔ Kustomization tkfg2 reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
✔ Kustomization tkfg3 reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
✔ Kustomization tkfg4 reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
✔ Kustomization tkfg5 reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
✔ Kustomization tkfg6 reconciliation completed
✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
```
The reconciliation success/error messages are printed in alphabetical order of the resource names.

closes #3746
closes #3793

### CODE REVIEW

**Comment by kingdonb** [comment]
This is awesome, we looked at your contribution yesterday at Flux Bug Scrub and we all agreed, it's a great idea 🎉 

Please give some time for maintainers to review the code.

It may also help if you can first rebase, now that the `2.0.0-rc.1` release has been cut.

**Comment by rishinair11** [comment]
> This is awesome, we looked at your contribution yesterday at Flux Bug Scrub and we all agreed, it's a great idea 🎉
> 
> Please give some time for maintainers to review the code.
> 
> It may also help if you can first rebase, now that the `2.0.0-rc.1` release has been cut.

Thank you so much for the acknowledgment :)
Yes, I'll rebase it now.

**Comment by rishinair11** [comment]
> The patching should be separated from the waiting so that we can have a patching phase and a waiting phase where the former would patch all objects and only then we wait for all of them to be reconciled. This is the behaviour that `kubectl` implements and it makes the overall process faster.

I guess the wait phase only occurs when we pass `--wait` right? So are you saying that when we pass `--wait`, we should wait after all the resources have been patched?

**Comment by rishinair11** [comment]
If yes, I feel it should be addressed in a separate issue? Is that in the scope of this card? Let me know, accordingly I will try to refactor.

**Comment by rishinair11** [comment]
@kingdonb @makkes 
If yes, I feel it should be addressed in a separate issue? Is that in the scope of this card? Let me know, accordingly I will try to refactor.

**Comment by kingdonb** [comment]
I think this is in scope here, it's not good to reinvent processes that are already established from projects that are well-known like kubectl. I haven't read the code in detail, but if I'm reading @makkes right, this wouldn't be accepted in its current form.

**Comment by rishinair11** [comment]
> I think this is in scope here, it's not good to reinvent processes that are already established from projects that are well-known like kubectl. I haven't read the code in detail, but if I'm reading @makkes right, this wouldn't be accepted in its current form.
> 
> A bit more concretely, if we do the workflow as patch-wait patch-wait patch-wait, then some waits may never finish because of the order of dependency. That's not only going to slow things down, it may cause the operation to fail, outright.

@kingdonb I totally understand this. Just to be clear, the patch-wait patch-wait patch-wait behaviour existed before my changes if you run:
```console
flux resume ks --all --wait
```
That's the reason why I was questioning this not being in the scope of this issue.

**Comment by rishinair11** [comment]
Sounds good, I'll try my best 😄

**Comment by makkes** [comment]
If this gets all too complex I'm happy to spin it out into its own PR at this point.

**Comment by rishinair11** [comment]
> If this gets all too complex I'm happy to spin it out into its own PR at this point.

@makkes I'd prefer it since it would change the behavior of `--wait` significantly. 
I shall create a new issue and start working on it meanwhile :)

**Comment by rishinair11** [comment]
Meanwhile, please let me know if changes/refactors are needed for this one :) 
Happy to collab.

**Comment by rishinair11** [comment]
On second thought, I would prefer to work on the "reconcile after resume" behavior in this PR. I will link this PR to both the issues I've created. 
This is because I'll have to wait for this PR to get merged before finalizing changes for the new `--wait` behavior.
Since the code is related, it will be easier for the reviewer as well.
Is that okay @makkes @kingdonb?

**Comment by rishinair11** [comment]
In this commit, I have changed the behavior for when `--wait` is passed during `flux resume ..`.
It now uses go-routines to reconcile each object concurrently.

There is just one potential issue I see with it is that the CLI output order of each object will be randomized (like if `ks1`, `ks2`, `ks3` are reconciled in that order, then their outputs might come as `ks3 reconciled`, `ks1 reconciled`, `ks2 reconciled`), which is fine, but it will make the e2e tests nondeterministic. 

```
=== RUN   TestKustomizationFromGit
    main_test.go:352: Mismatch from golden file 'testdata/kustomization/resume_kustomization_from_git_multiple_args_wait.golden': Mismatch from expected value (-want +got):
          (
                """
                ► resuming kustomization tkfg in tkfg-3 namespace
                ✔ kustomization resumed
        -       ◎ queued Kustomization for reconciliation
                ► resuming kustomization tkfg in tkfg-3 namespace
                ✔ kustomization resumed
        +       ◎ queued Kustomization for reconciliation
                ◎ queued Kustomization for reconciliation
                ✔ Kustomization tkfg reconciliation completed
                ... // 4 identical lines
                """
          )
        
    main_test.go:352: Mismatch from golden file 'testdata/kustomization/resume_kustomization_from_git_multiple_args_wait_ks_ready.golden': Mismatch from expected value (-want +got):
          (
                """
                ► resuming kustomization tkfg in tkfg-3 namespace
                ✔ kustomization resumed
        +       ► resuming kustomization tkfg in tkfg-3 namespace
        +       ✔ kustomization resumed
                ◎ queued Kustomization for reconciliation
        -       ✔ Kustomization tkfg reconciliation completed
        -       ✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
        -       ► resuming kustomization tkfg in tkfg-3 namespace
        -       ✔ kustomization resumed
                ◎ queued Kustomization for reconciliation
        +       ✔ Kustomization tkfg reconciliation completed
        +       ✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
                ✔ Kustomization tkfg reconciliation completed
                ✔ applied revision 6.3.5@sha1:67e2c98a60dc92283531412a9e604dd4bae005a9
                """
          )
        
--- FAIL: TestKustomizationFromGit (50.04s)
```

To avoid this, I can reconcile in a for loop without go-routines, we would still resume all resources first and reconcile later, but the reconciliation will always be blocking, thus slower.

I'd appreciate some insight into which approach can be taken.

**Comment by rishinair11** [comment]
AFAIK what happens is that:
- The object is patched with `.Spec.Suspend` set to `false`
- Once patched, we wait for `metav1.ConditionTrue` condition for the said object with a default timeout (or overridden by the user)

So you're right, technically the object `Spec` is patched and we are waiting for the respective controller to reconcile. But there doesn't seem to be a "second patch" as you said, or at least not in the code I am currently working with (`resume.go`) .

**Comment by rishinair11** [comment]
@kingdonb I have added this commit https://github.com/fluxcd/flux2/pull/3755/commits/23c3c4537f5cd5724ca520ab75b66096acf1a89c which basically makes `resume.reconcile()` return a response object and using that, I print out the success/error messages in a deterministic fashion. I have opted to print the objects in an alphabetical order based on their name. Let me know if that looks okay. 
Also, I am using a slice of `reconcileResponse` instead of a channel, as I need to sort it into alphabetical order, which lets me do away with an extra for-loop to transfer elements from the channel to the slice.

**Comment by kingdonb** [comment]
> dependsOn is only for Kustomizations

You're right. There is also `dependsOn` for HelmRelease, but this isn't important here and it doesn't change anything about what you say. I think printing in alphabetic order is a reasonable way to get the output to be deterministic.

Let's give some time for someone with write access to provide feedback; it may be a short while as most of the Flux devs are at KubeCon this week. Thanks for contributing this and also continuing to work with us to make it work most ideally!

**Comment by rishinair11** [comment]
Interesting. I didn't know that. Well, you learn something new everyday, just like I am learning about concurrency while working on this PR 😄 

I'll wait and maybe add minor tweaks to the code here and there until the Flux devs are back from KubeCon. 

Thanks @kingdonb ^·^

**Comment by rishinair11** [comment]
@kingdonb @makkes Any updates?

**Comment by makkes** [comment]
I'm on vacation this week. Will come back here next week when I'm back.

**Comment by kingdonb** [comment]
The 2.0.0-rc.4 release is out, can you please rebase @rishinair11 in case that can help further the review? 🙇 🙏

**Comment by kingdonb** [comment]
I'm excited about getting this merged. I reached for it the other day, as I thought it might have already been released.

I haven't tested this yet, but a glance at the PR shows the DCO check is failing. @rishinair11 This will have to be resolved, and perhaps another rebase, before this can be merged.

I'm unsure of whether this will go before or after the Flux 2.0.0 release (as it isn't a roadmap item, has been delayed for a while, I do suspect it will be after)

I've heard this release is imminent, and the planning is happening here:

* #4013

**Comment by makkes** [comment]
We want to get this into 2.0.0. I'll leave some comments later today or tomorrow latest.

**Comment by makkes** [comment]
@rishinair11 In order to get this in, we need your sign-off, though. See https://github.com/fluxcd/flux2/blob/main/CONTRIBUTING.md#certificate-of-origin for details.

**Comment by rishinair11** [comment]
@kingdonb @makkes Sorry, I failed to check my e-mail notifications on time haha. 
I have signed-off the commits. 
Please let me know if this is good-to-go or any other changes needed. 
Thank you for your excitement.

**Comment by makkes** [comment]
@rishinair11 I took the liberty of adding two commits that change the behaviour of resume/suspend just a little, mainly to maintain backwards-compatibility with older versions of the commands but also to improve the UX a little (e.g. to not process the same object multiple times).

The `--wait` flag for the `resume` command is now simpler to understand as well. If you're fine with these changes I'll squash all the commits and merge the PR.

**Comment by makkes** [comment]
@hiddeco @stefanprodan as I committed some code to this PR myself now, I'd appreciate another pair of eyes. 🙏🏻

**Comment by rishinair11** [comment]
@makkes Hey thanks for the contribution. Your changes makes sense to me and i'm glad theres a better UX for the --wait behaviour now. 

Looks like the `e2e / e2e-amd64-kubernetes (pull_request)` status is failing though.

**Comment by makkes** [comment]
> Looks like the [e2e / e2e-amd64-kubernetes (pull_request)](https://github.com/fluxcd/flux2/actions/runs/5404949152/jobs/9819830644#step:46:16) status is failing though.

Yeah, I fixed that one.

All commits squashed and ready for merge.

**Review by makkes — APPROVED** [review]

**Comment by rishinair11** [comment]
@makkes I have attempted to fix some golden files for the failing tests. I tested them on my local, and was able to reproduce as well as fix them. Could you run the CI again and check?

**Comment by rishinair11** [comment]
@makkes  Also, I noticed one thing, that when we run:
```console
❯ flux suspend ks robot-shop units
► suspending kustomization "robot-shop" in flux-system namespace
✔ kustomization suspended
► suspending kustomization "units" in flux-system namespace
✔ kustomization suspended
```
We have `"` around the Kustomization names.
But not during `resume`. Is that intentional, or should we add them there as well?

**Review by stefanprodan — CHANGES_REQUESTED** [review]
Please add examples to each command, how to suspend/resume multiple objects by name and how to do it for a whole namespace.

**Comment by rishinair11** [comment]
> Please add examples to each command, how to suspend/resume multiple objects by name and how to do it for a whole namespace.

Thanks for the review @stefanprodan. Could you point me to the appropriate place to put these examples?

**Comment by rishinair11** [comment]
Hey @stefanprodan 
Thanks for your review. 
I have added the requested examples.

```console
❯ flux suspend ks -h
The suspend command disables the reconciliation of a Kustomization resource.

Usage:
  flux suspend kustomization [name] [flags]

Aliases:
  kustomization, ks

Examples:
  # Suspend reconciliation for an existing Kustomization
  flux suspend ks podinfo

  # Suspend reconciliation for multiple Kustomizations
  flux suspend ks podinfo-1 podinfo-2

Flags:
  -h, --help   help for kustomization

Global Flags:
      --all                            suspend all resources in that namespace
...
...
```
It is done for all possible resumable and suspendible objects.

I have pushed a separate commit so it is easy to review the changes, please feel free to squash them before the eventual merge.

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @rishinair11 for this great improvement 🏅

**Comment by rishinair11** [comment]
Thank you so much @stefanprodan @makkes @kingdonb.
Great fun!

**Pull request merged**
